core/src/main/java/org/apache/brooklyn/entity/stock/ConditionalEntityImpl.java

```diff
@@ -19,25 +19,72 @@
+import org.apache.brooklyn.core.entity.Entities;
+import org.apache.brooklyn.core.entity.lifecycle.Lifecycle;
+import org.apache.brooklyn.core.entity.lifecycle.ServiceStateLogic;
+import org.apache.brooklyn.core.entity.trait.Startable;
+import org.apache.brooklyn.core.location.Locations;
+import org.apache.brooklyn.enricher.stock.Enrichers;
+import org.apache.brooklyn.util.collections.MutableSet;
+import org.apache.brooklyn.util.exceptions.Exceptions;
 
 public class ConditionalEntityImpl extends BasicStartableImpl implements ConditionalEntity {
 
+    private static final Logger LOG = LoggerFactory.getLogger(BasicStartableImpl.class);
+
     @Override
     public void start(Collection<? extends Location> locations) {
-        Entity child = sensors().get(CONDITIONAL_ENTITY);
-        EntitySpec<?> spec = config().get(CONDITIONAL_ENTITY_SPEC);
-        Boolean create = config().get(CREATE_CONDITIONAL_ENTITY);
-
-        // Child not yet created; Entity spec is present; Create flag is true
-        if (child == null && spec != null && Boolean.TRUE.equals(create)) {
-            Entity created = addChild(EntitySpec.create(spec));
-            sensors().set(CONDITIONAL_ENTITY, created);
+        try {
+            ServiceStateLogic.setExpectedState(this, Lifecycle.STARTING);
+
+            Entity child = sensors().get(CONDITIONAL_ENTITY);
+            EntitySpec<?> spec = config().get(CONDITIONAL_ENTITY_SPEC);
+            Boolean create = config().get(CREATE_CONDITIONAL_ENTITY);
+            Boolean propagate = config().get(PROPAGATE_CONDITIONAL_ENTITY_SENSORS);
+            Set<AttributeSensor<?>> sensors = MutableSet.copyOf(config().get(CONDITIONAL_ENTITY_SENSOR_LIST));
+
+            addLocations(locations);
+            locations = Locations.getLocationsCheckingAncestors(locations, this);
+            LOG.info("Starting entity "+this+" at "+locations);
+
+            // Child not yet created; Entity spec is present; Create flag is true
+            if (child == null && spec != null && Boolean.TRUE.equals(create)) {
+                child = addChild(EntitySpec.create(spec));
+                sensors().set(CONDITIONAL_ENTITY, child);
+
+                // Add enrichers for sensor propagateion
+                enrichers().add(Enrichers.builder().propagating(Startable.SERVICE_UP).from(child).build());
+                if (Boolean.TRUE.equals(propagate)) {
+                    if (sensors.isEmpty()) {
+                        enrichers().add(Enrichers.builder().propagatingAllButUsualAnd().from(child).build());
+                    } else {
+                        enrichers().add(Enrichers.builder().propagating(sensors).from(child).build());
+                    }
+                }
+            }
+
+            // Start child if create flag is set; otherwise just set service.isUp
+            if (Boolean.TRUE.equals(create)) {
+                Entities.invokeEffectorWithArgs(this, child, Startable.START, locations).getUnchecked();
```

Comment:
Why did you have to pass locations explicitly, won't it pick them from the parent by default? What if there's a location embedded in the spec?